### ISSUE TITLE
.NET Object instantiation via coreclr_create_delegate in C++ Runtime Host, no longer working with IUnknown/IDispatch in .NET Core 5.0

### ISSUE THREAD

**Issue by RickStrahl** [dotnet/runtime#48134]
This is a followup to issue #12904 in which I asked for help with runtime hosting the .NET Core CLR and instantiating a .NET object and passing it back as a COM object. This worked fine up until .NET Core 3.1 (just verified). 

However, recently with .NET Core 5.0 update, it appears that this approach no longer works. 

I use `coreclr_create_delegate` to create a reference to a .NET method and call it to return referenced .NET object. In .NET 5.0 this method still gets called and I can return non COM objects fine. But when I return a COM object I get an HR error: `Operation is not supported`. This same implementation works if I use the .NET Core 3.1 runtime.

Here's the code that I use to finalize the instantiation:

![image](https://user-images.githubusercontent.com/1374013/107584816-3cddaf00-6ba1-11eb-8c8f-c047e0cee07c.png)

Everything seems to work up to the point of returning the object result. If I return a string from the function - it works. If I return an anonymous object it works (but I can't cast it to IDispatch/IUnknonw). If I return my actual IUnknown casted objects the delegate call fails with `Operation is not supported`.

Here's the .NET Factory method and type definition that's being returned.

Factory:

```csharp
    public class wwDotnetBridgeFactory
    {
        // this is the method called from above - works with a simple result fails with an object
        public static int CreatewwDotnetBridgeByRef([MarshalAs(UnmanagedType.IUnknown)] ref object instance)
        {
            try
            {
                instance = new wwDotNetBridge();
            }
            catch (Exception ex)
            {
                instance = null;
                return ex.HResult;
            }

            return 0;
        }
    }
```

and the class header for the class returned:

```csharp
    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.AutoDual)]
    [ProgId("Westwind.wwDotNetBridge")]
    public class wwDotNetBridge
    { ... }
```

The C++ method delegate for the factory method:

```cpp
typedef HRESULT(__stdcall* createWwDotnetBridgeHandler)(CComPtr<IUnknown>* ptr);
createWwDotnetBridgeHandler createWwDotnetBridge;
```

The class is fairly simple - it just a function container with a long list of interop helper methods. There's no state and nothing funky about the class that wouldn't work over COM (and doesn't - I've been using it for this scenario for 10+ years).

**Some things I've tried**

* Using different `ClassInterface` settings (None, leaving out) - no change
* Explicitly casting to IDispatch - no change
* Leaving off the marshalling interfaces - no change
* Returning a simple .NET object (no COM attributes) - works but then can't cast to IDispatch

In all but the last case I get: **Operation is not Supported**

**Question: How do I return a COM accessible object instance now?**

> Bottom line is I need to be able to return a .NET object from the delegate and cast that object to an `IDispatch` instance somehow.

And yes I need a COM object here because I'm passing this object back to a calling Win32 application that uses COM to interact with .NET.

### Other Info
* This uses the 32 bit .NET Core runtime. 
* Works in .NET Core 3.1 - doesn't in .NET Core 5.0

**Comment by RickStrahl** [comment]
Thinking out loud here...

I wonder if it's possible to return a function pointer as an Integer to the C++ code. The C++ code could then wrap the IDispatch interface and return that to my client. In fact I think I could even pass the pointer all the way through to FoxPro, which can pick it up and turn it into an IDispatch instance. The ref counting would have to be figured out...

**Comment by AaronRobinsonMSFT** [comment]
@RickStrahl I don't see anything obviously wrong with your desired scenario. I will take a look at this.

One this that is a bit unusual is the signature - `typedef HRESULT(__stdcall* createWwDotnetBridgeHandler)(CComPtr<IUnknown>* ptr);` The argument type `CComPtr<IUnknown>*` is incorrect - this shouldn't impact anything I'm just pointing it out. The correct type is actually `IUnknown**` - see the `T** operator&()` override on `CComPtrBase<T>`. Pointers to `CComPtr<T>` instances are weird and I am unsure why they would ever be used.

**Comment by AaronRobinsonMSFT** [comment]
@RickStrahl I am not able to reproduce the error locally in the marshalling. I am doing something slightly different, marshalling a function pointer from managed instead of calling `coreclr_create_delegate()`, but the failing call logic should be the same. I verified this in both 32-bit and 64-bit in a local build of 6.0 and the official 5.0.2.

```C#
public unsafe class Program
{
    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.AutoDual)]
    [ProgId("Westwind.wwDotNetBridge")]
    public class wwDotNetBridge
    { }

    public static void CreateBridge([MarshalAs(UnmanagedType.IUnknown)] ref object instance)
    {
        instance = new wwDotNetBridge();
    }

    public delegate void Creator([MarshalAs(UnmanagedType.IUnknown)] ref object instance);

    static void Main(string[] args)
    {
        var f = new Creator(CreateBridge);
        Test_0(f);
    }

    [DllImport(@"NativeLibrary.dll", EntryPoint = "Test")]
    static extern void Test_0(Creator fptr);
}
```

``` C++
#define EXPORT_FUNC __declspec(dllexport)
#define CALLCONV __stdcall
extern "C"
{
    using fptr_t = void(CALLCONV*)(_Outptr_ IUnknown** ptr);

    EXPORT_FUNC void CALLCONV Test(fptr_t fptr)
    {
        IUnknown* pUnk = nullptr;
        fptr(&pUnk);
        pUnk->Release();
    }
}
```

**Comment by RickStrahl** [comment]
Aaron - thank you so much for your patience.

It turns out that I've completely misdiagnosed this problem. The problem is with the object that I'm trying to instantiate, which works in 3.1 but not 5.0. It took me a bit of experimenting to find the problem  because I did not expect a failure in the object instantiation CTOR of this simple object.

The issue turns out to be a deprecated value in .NET 5.0 with `SecurityProtocol`  which was set in the CTOR (this is a legacy tool so all sorts of old code and versions).

```text
2/23/2021 9:10:52 PM - Incoming wwDotnetBridge creation.
2/23/2021 9:10:52 PM - ERROR: The requested security protocol is not supported..   <---  
```

Turns out the object I'm creating is using the following in the CTOR():

```cs
if (!ServicePointManager.SecurityProtocol.HasFlag(SecurityProtocolType.Tls12))
{
    // support for all SSL/TLS protocols
    ServicePointManager.SecurityProtocol =
        SecurityProtocolType.Tls11 |
        SecurityProtocolType.Tls12 |
        SecurityProtocolType.Ssl3 |
        SecurityProtocolType.Tls;
}
```
SSL3 is no longer supported in 5.0 and this is what broke the code. I finally figured this out when I placed a logging exception handler around the factory method (which I should have done from the get go).  IAC, after fixing the CTOR of the class, I now get the runtime to instantiate using the existing code I had.

![image](https://user-images.githubusercontent.com/1374013/108963375-1b0f0e00-761e-11eb-8178-779bd5a45076.png)

Also thank you for the nudge about the ancient ATL COM pointers. My C++ code is extremely rusty, and COM code even rustier so I used what's already been in use. Trying the different pointer types - even though they had no effect - ended up helping me figure out the problem in the component load rather than the code. 


One last thing of interest here - I'm actually instantiating a full framework assembly in .NET Core. Before I found the CTOR issue I noticed that I tried to throw up a messagebox and that would just bomb out the code bypassing the exception handlers in .NET and even in C++ and throwing a native error in the host application. Removing the Framework specific code allowed me to get through the method at least and log. 

All the component does is basically provide a bunch of Reflection helpers  to surface functionality that COM Interop doesn't support and so far I haven't seen any issues. It works, but of course it'll blow at runtime when something isn't supported.  Still it's pretty cool that this works at all (ie. using a Full Framework Assembly in Core).

Anyway - thank again for getting me on the right track Aaron!
